symengine/subs.h

```diff
@@ -276,6 +276,238 @@ class SubsVisitor : public BaseVisitor<SubsVisitor, XReplaceVisitor>
+                    }
+                } else
+                    d = dict;
+            } else if (is_a<Pow>(*sub1)) {
+                RCP<const Pow> subst = rcp_static_cast<const Pow>(sub1);
+                auto sub1_exp = subst->get_exp();
+                auto sub1_base = subst->get_base();
+                exists = false;
+                if (is_a_Number(*sub1_exp)) {
+                    for (const auto &p : dict) {
+                        auto diff_ = sub(p.second, sub1_exp);
+                        if (eq(*sub1_base, *(p.first))
+                            and eq(*sub1_exp, *p.second)) {
+                            exists = true;
+                        } else if (eq(*sub1_base, *(p.first))
+                                   and down_cast<const Number &>(*diff_)
+                                           .is_positive()) {
+                            exists = true;
+                            Mul::dict_add_term_new(outArg(coef), d,
+                                                   sub(p.second, sub1_exp),
+                                                   p.first);
+                        } else {
+                            Mul::dict_add_term_new(outArg(coef), d, p.second,
+                                                   p.first);
+                        }
+                    }
+                } else {
+                    for (const auto &p : dict) {
+                        if (eq(*sub1_base, *(p.first))
+                            and eq(*sub1_exp, *p.second)) {
+                            exists = true;
+                        } else {
+                            Mul::dict_add_term_new(outArg(coef), d, p.second,
+                                                   p.first);
+                        }
+                    }
+                }
+                if (exists) {
+                    if (is_a_Number(*rep)) {
+                        if (down_cast<const Number &>(*rep).is_zero()) {
+                            result_ = rep;
+                            return;
+                        }
+                        imulnum(outArg(coef),
+                                rcp_static_cast<const Number>(rep));
+                    } else if (is_a<Mul>(*rep)) {
+                        RCP<const Mul> tmp = rcp_static_cast<const Mul>(rep);
+                        imulnum(outArg(coef), tmp->get_coef());
+                        for (const auto &q : tmp->get_dict()) {
+                            Mul::dict_add_term_new(outArg(coef), d, q.second,
+                                                   q.first);
+                        }
+                    } else {
+                        RCP<const Basic> exp, t;
+                        Mul::as_base_exp(rep, outArg(exp), outArg(t));
+                        Mul::dict_add_term_new(outArg(coef), d, exp, t);
+                    }
+                } else
+                    d = x.get_dict();
+            } else if (is_a<Symbol>(*sub1)) {
```

Comment:
Can you explain this part? Why is `sub1` being a Symbol important? I thought the issue was that `sub1` being a Mul isn't handled correctly.

Reply:
Partial substitution can happen for `Symbol` objects as well. For example, if the expression is `x**2*y` and `sub1` is `x`, there is still partial substitution possible. 
The first part only looks for exact matches.

Follow up question, should substitution happen repeatedly until not possible? For instance, if the expression is `x**2*y**3` and `x*y` is to be replaced by `z`, should the result be `x*y**2*z` or `y*z**2`?

Reply:
Hmm, `(x**2*y).subs({x: z})` didn't work before?

Reply:
It did, but I've made the non `fast_exec` implementation complete. Should I remove these redundant parts?

Reply:
ping @isuruf

Reply:
If it's redundant, remove them